google_analytics/aggregation.py

```diff
@@ -1,7 +1,138 @@
+                             nrows=nrows,
+                             chunksize=chunksize):
+
+        chunk = chunk.reset_index()
+
+        # Normalize JSON columns
+        for column in JSON_COLUMNS:
+            column_as_df = pd.io.json.json_normalize(chunk[column])
+            chunk = chunk.drop(column, axis=1).merge(column_as_df, right_index=True, left_index=True)
+
+        # Parse date
+        chunk['date'] = chunk['date'].apply(lambda x: pd.datetime.strptime(str(x), '%Y%m%d'))
+
+        # Only keep relevant columns
+        cols = ['date', 'fullVisitorId', 'operatingSystem', 'country', 'browser',
+                'pageviews', 'transactions', 'visits', 'transactionRevenue']
+        try:
+            chunk = chunk[cols]
+        except KeyError as e:
+            # Regex magic to find exactly which columns were not found.
+            # Might be different in Python 3, be careful!
+            missing_cols = list(re.findall(r"'(.*?)'", e.args[0]))
+            for col in missing_cols:
+                print("Column {} was not found in chunk {}, filling with zeroes".format(col, i))
+                chunk[col] = [0] * len(chunk)
+            chunk = chunk[cols]
+
+        print("Loaded chunk {}, Shape is: {}".format(i, chunk.shape))
+        chunk.to_csv(os.path.join(temp_dir, str(i) + ".csv"), encoding='utf-8', index=False)
+        i += 1
+
+    print("Finished all chunks, now concatenating")
+    files = glob.glob(os.path.join(temp_dir, "*.csv"))
+    with open(output, 'wb') as outfile:
+        for i, fname in enumerate(files):
+            with open(fname, 'rb') as infile:
+                # Throw away header on all but first file
+                if i != 0:
+                    infile.readline()
+                    # Block copy rest of file from input to output without parsing
+                shutil.copyfileobj(infile, outfile)
+
+
+def aggregate(df):
+    """Group and pivot the dataframe so that we have one row per visitor.
+
+    This row includes features of two types:
+      * Dynamic. These are repeated for each month and capture the time-series like behavior.
+                 Some examples are the revenue and visits per month, for every month in the dataset.
+
+      * Static. These exist once per user and correspond to relatively constant properties.
+                Examples include the person's country, OS and browser.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        The original reduced dataframe - see the `reduce_df` function
+
+    Returns
+    -------
+    pd.DataFrame
+        The same dataframe grouped and pivoted, where each visitor is a single row.
+
+    """
+    df['transactions'].fillna(0, inplace=True)
+    df['transactionRevenue'].fillna(0.0, inplace=True)
+
+    def agg_date(s):
+        date = pd.datetime.strptime(s, '%Y-%m-%d')
+        return "{}_{}".format(date.month, date.year)
+
+    df["date"] = df["date"].apply(agg_date)
+
+    def most_frequent(x):
```

Comment:
What is the benefit of only including the most frequent value? In the end we still need to apply OHE right, so we still end up with a column for each country.

Reply:
actually...you are right :P For the countyr it does because I was thinking of replacing it with the financial metrics but for the rest maybe I should OHE directly.